File: Python Language/ProblemID_53.py
# https://leetcode.com/problems/maximum-subarray/submissions/

"""
Problem Description

Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.
"""

# Time Complexity   Space Complexity
#      O(N)              O(1)

import sys
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        maximum = - sys.maxsize - 1
        max_value = 0
        for each_num in nums:
            max_value+=each_num
            if(max_value>maximum):
                maximum = max_value
            if(max_value<0):
                max_value = 0
        return maximum
        
        

File: C++/ProblemID_53.cpp
// https://leetcode.com/problems/maximum-subarray/

/*

Problem Description 

Given an integer array nums, find the contiguous subarray (containing at least one number) which has the largest sum and return its sum.

A subarray is a contiguous part of an array.


Time Complexity    Space Complexity
    O(N)             O(1)

*/

#include <iostream>
#include <climits>
#include <vector>

using namespace std;

class Solution
{
public:
  int maxSubArray(vector<int> &nums)
  {
    int cs = 0;
    int ms = INT_MIN;
    for (auto &i : nums)
    {
      cs += i;
      ms = max(ms, cs);
      if (cs < 0)
        cs = 0;
    }
    return ms;
  }
};

int main()
{
  vector<int> v = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
  Solution solution;
  cout << solution.maxSubArray(v) << '\n';
  return 0;
}